Content.Server/Lightning/LightningSystem.cs

```diff
@@ -69,70 +49,53 @@ private void OnCollide(EntityUid uid, LightningComponent component, ref StartCol
     /// <param name="lightningPrototype">The prototype for the lightning to be created</param>
     public void ShootLightning(EntityUid user, EntityUid target, string lightningPrototype = "Lightning")
     {
+        if (Deleted(user) || Deleted(target))
+            return;
+
         var spriteState = LightningRandomizer();
         _beam.TryCreateBeam(user, target, lightningPrototype, spriteState);
+
+        var ev = new HitByLightningEvent(user, target);
+        RaiseLocalEvent(target, ref ev, true);
     }
 
     /// <summary>
-    /// Looks for a target to arc to in all 8 directions, adds the closest to a local dictionary and picks at random
+    /// Looks for objects with a LightningTarget component in the radius, prioritizes them, and hits the highest priority targets with lightning.
     /// </summary>
-    /// <param name="component"></param>
-    /// <param name="target"></param>
-    /// <param name="controllerEntity"></param>
-    private void Arc(LightningComponent component, EntityUid target, EntityUid controllerEntity)
+    /// <param name="user">Where the lightning fires from</param>
+    /// <param name="range">Targets selection radius</param>
+    /// <param name="boltCount">Number of lightning bolts</param>
+    /// <param name="lightningPrototype">The prototype for the lightning to be created</param>
+    /// <param name="arcDepth">how many times to recursively fire lightning bolts from the target points of the first shot.</param>
+    public void ShootRandomLightnings(EntityUid user, float range, int boltCount, string lightningPrototype = "Lightning", int arcDepth = 0)
     {
-        if (!TryComp<BeamComponent>(controllerEntity, out var controller))
-            return;
-
-        var targetXForm = Transform(target);
-        var directions = Enum.GetValues<Direction>().Length;
+        //To Do: add support to different priority target tablem for different lightning types
+        //To Do: Remove Hardcode LightningTargetComponent (this should be a parameter of the SharedLightningComponent)
+        var targets = _lookup.GetComponentsInRange<LightningTargetComponent>(Transform(user).MapPosition, range).ToList(); //To Do: remove hardcode component
+        _random.Shuffle(targets);
```

Comment:
This is really expensive, don't re-shuffle and keep a list around instead.

Reply:
This is important! If not mixed, the lightning bolts will always strike the same target within range

Reply:
while keeping the targets sorted by priority

Reply:
![image](https://github.com/space-wizards/space-station-14/assets/96445749/b8b046e4-3783-4233-bd74-aae28d2cbb52)
the system sorts all targets in the radius by priority, shuffles them within its priority, and hits them in the order of that list. The system sorts all targets in the radius by priority, shuffles them within its priority, and hits them in the order of the list.

Reply:
For every lightning it's going to re-allocate a new list, then shuffle it, then sort it, and doing all 3 is going to allocate a lot and be very slow.

Reply:
If there are suggestions on how to change this to avoid linq, I'm listening.